lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs

```diff
@@ -175,22 +177,22 @@ improveTxOut (opt0, selection, utxo0) (inps0, txout) = do
     target = mkTargetRange txout
 
     improve
-        :: forall m e. MonadRandom m
-        => (CoinSelectionOptions e, [(TxIn, TxOut)], UTxO)
-        -> m (CoinSelectionOptions e, [(TxIn, TxOut)], UTxO)
-    improve (opt@(CoinSelectionOptions maxN fn), inps, utxo)
+        :: forall m. MonadRandom m
+        => (Word64, [(TxIn, TxOut)], UTxO)
+        -> m (Word64, [(TxIn, TxOut)], UTxO)
+    improve (maxN, inps, utxo)
         | maxN >= 1 && balance' inps < targetAim target = do
             runMaybeT (pickRandomT utxo) >>= \case
                 Nothing ->
-                    return (opt, inps, utxo)
+                    return (maxN, inps, utxo)
                 Just (io, utxo') | isImprovement io inps -> do
                     let inps' = io : inps
-                    let opt' = CoinSelectionOptions (maxN - 1) fn
-                    improve (opt', inps', utxo')
+                    let maxN' = maxN - 1
+                    improve (maxN', inps', utxo')
                 Just _ ->
-                    return (opt, inps, utxo)
+                    return (maxN, inps, utxo)
         | otherwise =
-            return (opt, inps, utxo)
+            return (maxN, inps, utxo)
 
```

Comment:
Note that most of the changes in this file are because the type for the `maxN` in the coin selection options was changed from `Word64` to `Word8 -> Word8`. Therefore, the actual `maxN` is now computed in the root `random` (or `largestFirst` function, and passed down to the iteration function as a `Word64` to be depleted.